src/mono/mono/mini/mini-s390x.c

```diff
@@ -2212,6 +2659,227 @@ mono_arch_lowering_pass (MonoCompile *cfg, MonoBasicBlock *bb)
+                       		case CMP_GE_UN:
+                       		        temp = ins->sreg1;
+                       		        ins->sreg1 = ins->sreg2;
+                       		        ins->sreg2 = temp;
+                       		case CMP_LE:
+                       		case CMP_LE_UN:{
+                       		        NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_gt_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                       		        NEW_SIMD_INS (cfg, ins, temp_ins, OP_S390_VNO, s390_vr16, s390_vr16, s390_vr16);
+                       		        break;
+                       		}
+                       		default:
+                       		        g_assert_not_reached ();
+                       		        break;
+                        	}
+				break;
+			}
+			case OP_XCOMPARE_FP:{
+                	        switch (ins->inst_c0 >> 4){
+                	        case CMP_EQ:
+					NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_comp_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                	                break;
+                	        case CMP_LT_UN:
+                	        case CMP_LT:
+                	                temp = ins->sreg1;
+                	                ins->sreg1 = ins->sreg2;
+                	                ins->sreg2 = temp;
+                	        case CMP_GT_UN:
+                	        case CMP_GT:
+					NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_gt_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                	                break;
+                	        case CMP_LE_UN:
+                	        case CMP_LE:
+                	                temp = ins->sreg1;
+                	                ins->sreg1 = ins->sreg2;
+                	                ins->sreg2 = temp;
+                	        case CMP_GE_UN:
+                	        case CMP_GE:
+					NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_ge_fp_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                	                break;
+                	        default:
+                	                g_assert_not_reached ();
+                	                break;
+                	        }
+			}
+				break;
+			default:
+				g_assert_not_reached();
+				break;
+		}
+		switch (type_is_float(GTMREG_TO_INT(ins->inst_c1))){
+		case OP_XCOMPARE:
+			ins->opcode = GINT_TO_OPCODE (simd_type_to_extract_int_op (GTMREG_TO_INT (ins->inst_c0 & 0x0f), GTMREG_TO_INT (ins->inst_c0 >> 4)));
+			break;
+		case OP_XCOMPARE_FP:
+			ins->opcode = GINT_TO_OPCODE (simd_type_to_extract_fp_op (GTMREG_TO_INT (ins->inst_c0 & 0x0f), GTMREG_TO_INT (ins->inst_c0 >> 4)));
+			break;
+		default:
+			g_assert_not_reached ();
+			break;
+		}
+		/* we don't use a register rather the CC set by the vector compare instructions */
+		ins->sreg1 = -1;
```

Comment:
We'd now have to invalidate `sreg2` too, right?   Or mabye it would be cleaner to emit both instructions via `NEW_SIMD_INS` and the nullify the combined instruction?